tao_compiler/mlir/disc/transforms/disc_cse.cc

```diff
@@ -0,0 +1,213 @@
+      llvm::sort(lhsOperandStorage, [](Value a, Value b) -> bool {
+        return a.getAsOpaquePointer() < b.getAsOpaquePointer();
+      });
+      lhsOperands = lhsOperandStorage;
+
+      rhsOperandStorage.append(rhsOperands.begin(), rhsOperands.end());
+      llvm::sort(rhsOperandStorage, [](Value a, Value b) -> bool {
+        return a.getAsOpaquePointer() < b.getAsOpaquePointer();
+      });
+      rhsOperands = rhsOperandStorage;
+    }
+    auto checkValueRangeMapping =
+        [](ValueRange lhs, ValueRange rhs,
+           function_ref<LogicalResult(Value, Value)> mapValues) {
+          for (auto operandPair : llvm::zip(lhs, rhs)) {
+            Value curArg = std::get<0>(operandPair);
+            Value otherArg = std::get<1>(operandPair);
+            if (curArg.getType() != otherArg.getType()) return false;
+            if (failed(mapValues(curArg, otherArg))) return false;
+          }
+          return true;
+        };
+    // Check mapping of operands and results.
+    if (!checkValueRangeMapping(lhsOperands, rhsOperands, mapOperands))
+      return false;
+    if (!checkValueRangeMapping(lhs->getResults(), rhs->getResults(),
+                                mapResults))
+      return false;
+
+    auto lhsReduceOp = dyn_cast<mhlo::ReduceOp>(lhs);
+    auto rhsReduceOp = dyn_cast<mhlo::ReduceOp>(rhs);
+    if (lhsReduceOp && rhsReduceOp) {
+      auto& lhsBody = lhsReduceOp.getBody();
+      auto& rhsBody = rhsReduceOp.getBody();
+      if (not(lhsBody.hasOneBlock() && rhsBody.hasOneBlock())) return false;
+
+      for (auto it : llvm::zip(lhsBody.front().getOperations(),
+                               rhsBody.front().getOperations())) {
+        auto& lhsOp = std::get<0>(it);
+        auto& rhsOp = std::get<1>(it);
+        // we currently do not compare equality of nested regions
+        if (lhsOp.getNumRegions() != 0 or rhsOp.getNumRegions() != 0)
+          return false;
+
+        auto innerOpMatch = OperationEquivalence::isEquivalentTo(
+            &lhsOp, &rhsOp,
+            /*mapOperands=*/OperationEquivalence::ignoreValueEquivalence,
+            /*mapResults=*/OperationEquivalence::ignoreValueEquivalence,
+            OperationEquivalence::IgnoreLocations);
+        if (not innerOpMatch) return false;
+      }
+      return true;
+    }
+    if (lhs->getNumRegions() != 0 or rhs->getNumRegions() != 0) return false;
+
+    return true;
+  }
+
+  bool operator()(Operation* const& lhsC, Operation* const& rhsC) const {
+    auto* lhs = const_cast<Operation*>(lhsC);
+    auto* rhs = const_cast<Operation*>(rhsC);
+
+    if (lhs == nullptr or rhs == nullptr) return false;
+    return isEquivalentTo(
+        lhs, rhs,
+        /*mapOperands=*/OperationEquivalence::exactValueMatch,
+        /*mapResults=*/OperationEquivalence::ignoreValueEquivalence,
+        OperationEquivalence::IgnoreLocations);
+  }
+};
+}  // namespace
+
+namespace {
+struct DiscMhloCSEPass : public DiscMhloCSEPassBase<DiscMhloCSEPass> {
```

Comment:
can you provide more messages that why we need disc-cse  pass, instead of llvm-ces?

Reply:
The original one provided in LLVM will not work on operations with body regions such as `mhlo::ReduceOp`. The PR adds CSE for `mhlo::ReduceOp` and creates the interface to add more other mhlo operations in the future.